File: 08-greedy/examples.py
"""
Жадные алгоритмы: классические задачи.
Запуск: python examples.py
"""

import heapq


# ── Activity selection ───────────────────────────────────────────────
def activity_selection(intervals: list[tuple[int, int]]) -> list[tuple[int, int]]:
    """Выбираем максимум непересекающихся интервалов.

    Критерий: брать тот, что заканчивается раньше.
    """
    # сортируем по времени окончания
    sorted_iv = sorted(intervals, key=lambda x: x[1])
    chosen: list[tuple[int, int]] = []
    last_end = float("-inf")
    for s, f in sorted_iv:
        if s >= last_end:
            chosen.append((s, f))
            last_end = f
    return chosen


# ── Размен монет (канонические номиналы) ─────────────────────────────
def coin_change_greedy(coins: list[int], amount: int) -> list[int] | None:
    """Работает только для канонической системы (1, 5, 10, 50, 100, ...).

    Для произвольных монет нужно ДП — см. 07-dynamic-programming.
    """
    coins_desc = sorted(coins, reverse=True)
    result = []
    for c in coins_desc:
        while amount >= c:
            amount -= c
            result.append(c)
    return result if amount == 0 else None


# ── Алгоритм Хаффмана (генерация префиксного кода) ───────────────────
def huffman_codes(freq: dict[str, int]) -> dict[str, str]:
    """Возвращает кодировку: символ → битовая строка."""
    # heap из (частота, уникальный_id, поддерево)
    counter = 0
    heap: list[tuple[int, int, object]] = []
    for ch, f in freq.items():
        heapq.heappush(heap, (f, counter, ch))
        counter += 1

    while len(heap) > 1:
        f1, _, left = heapq.heappop(heap)
        f2, _, right = heapq.heappop(heap)
        heapq.heappush(heap, (f1 + f2, counter, (left, right)))
        counter += 1

    codes: dict[str, str] = {}

    def assign(node, prefix: str) -> None:
        if isinstance(node, str):
            codes[node] = prefix or "0"  # на случай 1 символа
            return
        left, right = node
        assign(left, prefix + "0")
        assign(right, prefix + "1")

    if heap:
        assign(heap[0][2], "")
    return codes


# ── Заправки на трассе ───────────────────────────────────────────────
def min_refuels(stations: list[int], distance: int, tank: int) -> int:
    """Сколько раз минимум придётся заправиться, чтобы проехать distance км.

    stations отсортированы по возрастанию, бак заполнен в начале.
    """
    points = stations + [distance]
    refuels = 0
    current_pos = 0
    last_refuel = 0
    i = 0
    while current_pos + tank < distance:
        # доезжаем до самой дальней досягаемой
        farthest = -1
        while i < len(points) and points[i] <= current_pos + tank:
            farthest = points[i]
            i += 1
        if farthest == -1 or farthest == last_refuel:
            return -1  # не дотягиваем
        current_pos = farthest
        last_refuel = farthest
        refuels += 1
        if current_pos == distance:
            return refuels
    return refuels


if __name__ == "__main__":
    iv = [(1, 4), (3, 5), (0, 6), (5, 7), (3, 9), (5, 9), (6, 10), (8, 11), (8, 12), (2, 14), (12, 16)]
    print("activity_selection:", activity_selection(iv))

    print("coin_change_greedy [1,5,10,25,100], 187:", coin_change_greedy([1, 5, 10, 25, 100], 187))
    print("coin_change_greedy [1,3,4], 6 (тут жадность даёт 3, а оптимум 2!):",
          coin_change_greedy([1, 3, 4], 6))

    print("huffman:", huffman_codes({"a": 5, "b": 9, "c": 12, "d": 13, "e": 16, "f": 45}))

    print("min_refuels:", min_refuels([200, 375, 550, 750, 950], distance=1000, tank=400))

File: 08-greedy/examples.go
// Жадные алгоритмы на Go.
// Запуск: go run examples.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

// ── Activity selection ─────────────────────────────────────────────
type Interval struct{ Start, Finish int }

func ActivitySelection(intervals []Interval) []Interval {
	sorted := make([]Interval, len(intervals))
	copy(sorted, intervals)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Finish < sorted[j].Finish })

	chosen := []Interval{}
	lastEnd := -1 << 30
	for _, iv := range sorted {
		if iv.Start >= lastEnd {
			chosen = append(chosen, iv)
			lastEnd = iv.Finish
		}
	}
	return chosen
}

// ── Размен монет (канонические номиналы) ───────────────────────────
func CoinChangeGreedy(coins []int, amount int) []int {
	desc := make([]int, len(coins))
	copy(desc, coins)
	sort.Sort(sort.Reverse(sort.IntSlice(desc)))

	res := []int{}
	for _, c := range desc {
		for amount >= c {
			amount -= c
			res = append(res, c)
		}
	}
	if amount != 0 {
		return nil
	}
	return res
}

// ── Хаффман: построение кода через min-heap ────────────────────────
type huffNode struct {
	freq        int
	symbol      byte // 0 если внутренний узел
	left, right *huffNode
}

type huffHeap []*huffNode

func (h huffHeap) Len() int            { return len(h) }
func (h huffHeap) Less(i, j int) bool  { return h[i].freq < h[j].freq }
func (h huffHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *huffHeap) Push(x any)         { *h = append(*h, x.(*huffNode)) }
func (h *huffHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func HuffmanCodes(freq map[byte]int) map[byte]string {
	h := &huffHeap{}
	heap.Init(h)
	for sym, f := range freq {
		heap.Push(h, &huffNode{freq: f, symbol: sym})
	}
	for h.Len() > 1 {
		a := heap.Pop(h).(*huffNode)
		b := heap.Pop(h).(*huffNode)
		heap.Push(h, &huffNode{freq: a.freq + b.freq, left: a, right: b})
	}

	codes := map[byte]string{}
	if h.Len() == 0 {
		return codes
	}
	var walk func(n *huffNode, prefix string)
	walk = func(n *huffNode, prefix string) {
		if n.left == nil && n.right == nil {
			if prefix == "" {
				prefix = "0"
			}
			codes[n.symbol] = prefix
			return
		}
		walk(n.left, prefix+"0")
		walk(n.right, prefix+"1")
	}
	walk(heap.Pop(h).(*huffNode), "")
	return codes
}

func main() {
	iv := []Interval{
		{1, 4}, {3, 5}, {0, 6}, {5, 7}, {3, 9},
		{5, 9}, {6, 10}, {8, 11}, {8, 12}, {2, 14}, {12, 16},
	}
	fmt.Println("ActivitySelection:", ActivitySelection(iv))
	fmt.Println("CoinChangeGreedy [1,5,10,25,100], 187:", CoinChangeGreedy([]int{1, 5, 10, 25, 100}, 187))

	codes := HuffmanCodes(map[byte]int{'a': 5, 'b': 9, 'c': 12, 'd': 13, 'e': 16, 'f': 45})
	for sym, code := range codes {
		fmt.Printf("  %c -> %s\n", sym, code)
	}
}
